esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to get ConfigurationDirectory: %s.", groupName.str());
+
+    SocketEndpointArray epa;
+    ForEachItemIn(i, locations)
+    {
+        SocketEndpoint ep(locations.item(i));
+        if (ep.isNull())
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid location '%s'.", locations.item(i));
+        epa.append(ep);
+    }
+    group.setown(createIGroup(epa));
+    queryNamedGroupStore().add(groupName.str(), group, false, defaultDir.str(), (clusterType == HThorCluster) ? grp_hthor : grp_thor);
+    ESPLOG(LogMin, "DFUFileIGroup %s added", groupName.str());
+    return group.getClear();
+}
+
+void CWsDfuEx::exportRecordDefinitionBinaryType(const char *recordDefinition, MemoryBuffer &layoutBin)
+{
+    MultiErrorReceiver errs;
+    Owned<IHqlExpression> expr = parseQuery(recordDefinition, &errs);
+    if (errs.errCount() > 0)
+    {
+        StringBuffer errorMsg;
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in parsing ECL %s: %s.", recordDefinition, errs.toString(errorMsg).str());
+    }
+    if (!expr)
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in parsing ECL: %s.", recordDefinition);
+    
+    if (!exportBinaryType(layoutBin, expr, false))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in exportBinaryType.");
+}
+
+IFileDescriptor *CWsDfuEx::createDFUDescriptor(IEspContext &context, const char *fileName, ClusterType clusterType,
+    const char *jobId, const char *userId, StringArray &locations, StringBuffer &groupName)
+{
+    Owned<IGroup> group = queryDFUFileIGroup(groupName, clusterType, locations);
+
+    Owned<IFileDescriptor> fileDesc = createFileDescriptor();
+    fileDesc->queryProperties().setProp("@job", jobId);
+    if (!isEmptyString(userId))
+        fileDesc->queryProperties().setProp("@owner", userId);
+
+    StringBuffer curDir, defaultDir;
+    if (!getConfigurationDirectory(nullptr, "data", (clusterType == HThorCluster) ? "hthor" : "thor", groupName, defaultDir))
+        makePhysicalPartName(fileName, 0, 0, curDir, false, DFD_OSdefault); // legacy
+    else
+        makePhysicalPartName(fileName, 0, 0, curDir, false, SepCharBaseOs(getPathSepChar(defaultDir)), defaultDir.str());
+    fileDesc->setDefaultDir(curDir.str());
+
+    StringBuffer partMask;
+    unsigned parts = group->ordinality();
+    getPartMask(partMask, fileName, parts);
+    fileDesc->setNumParts(parts);
+    fileDesc->setPartMask(partMask);
+
+    ClusterPartDiskMapSpec mspec;
+    mspec.defaultCopies = DFD_DefaultCopies;
+    fileDesc->addCluster(groupName.str(), group, mspec);
+
+    return fileDesc.getClear();
+}
+
+void CWsDfuEx::getFileAccessBeforePublish(IEspContext &context, const char *fileName, const char *jobId,
+    MemoryBuffer& layoutBin, Owned<IFileDescriptor> &fileDesc, IUserDescriptor *udesc,
```

Comment:
why is this a Owned<IFileDescriptor> &
it implies something is going to alter it..

Should use base type, I think should be change to const IFileDescriptor &

Reply:
ok